src/mds/CInode.cc

```diff
@@ -4391,35 +4391,87 @@ int64_t CInode::get_backtrace_pool() const
   }
 }
 
+class C_CInode_AuxSubtree : public MDSInternalContext {
+public:
+  explicit C_CInode_AuxSubtree(CInode *in) : MDSInternalContext(in->mdcache->mds), in(in) {
+    in->get(MDSCacheObject::PIN_PTRWAITER);
+  }
+  ~C_CInode_AuxSubtree() {
+    in->put(MDSCacheObject::PIN_PTRWAITER);
+  }
+
+  void finish(int r) override {
+    in->maybe_export_pin();
+  }
+private:
+  CInode *in;
+};
+
 void CInode::maybe_export_pin()
 {
-  if (g_conf->mds_bal_export_pin && is_dir() && !mdcache->export_pin_queue.count(this)) {
+  if (g_conf->mds_bal_export_pin && is_dir()) {
     dout(20) << "maybe_export_pin " << *this << dendl;
     mds_rank_t pin = get_projected_inode()->export_pin;
     if (pin == MDS_RANK_NONE) {
-      /* Try to find farthest full auth parent fragment which is pinned
-       * elsewhere.  There cannot be a break in the authority chain of
-       * directories, otherwise this inode itself will not be exported.
-       */
-      CInode *auth_last = this; /* N.B. we may not be full auth for any fragments of this inode, but adding it to the queue is harmless. */
-      bool auth_barrier = false;
-      for (CDir *cd = get_projected_parent_dir(); cd && !cd->inode->is_base() && !cd->inode->is_system(); cd = cd->inode->get_projected_parent_dir()) {
-        if (cd->is_full_dir_auth() && !auth_barrier) {
-          auth_last = cd->inode;
-        } else {
-          auth_barrier = true;
-        }
-        pin = cd->inode->get_projected_inode()->export_pin;
-        if (pin != MDS_RANK_NONE) {
-          if (pin != mdcache->mds->get_nodeid()) {
-            dout(20) << "adding ancestor to export_pin_queue " << *auth_last << dendl;
-            mdcache->export_pin_queue.insert(auth_last);
+      if (!mdcache->export_pin_queue.count(this)) {
+        /* Try to find farthest full auth parent fragment which is pinned
+         * elsewhere.  There cannot be a break in the authority chain of
+         * directories, otherwise this inode itself will not be exported.
+         */
+        CInode *auth_last = this; /* N.B. we may not be full auth for any fragments of this inode, but adding it to the queue is harmless. */
+        bool auth_barrier = false;
+        for (CDir *cd = get_projected_parent_dir(); cd && !cd->inode->is_base() && !cd->inode->is_system(); cd = cd->inode->get_projected_parent_dir()) {
+          if (cd->is_full_dir_auth() && !auth_barrier) {
+            auth_last = cd->inode;
+          } else {
+            auth_barrier = true;
+          }
+          pin = cd->inode->get_projected_inode()->export_pin;
+          if (pin != MDS_RANK_NONE) {
+            if (pin != mdcache->mds->get_nodeid()) {
+              dout(20) << "adding ancestor to export_pin_queue " << *auth_last << dendl;
+              mdcache->export_pin_queue.insert(auth_last);
+            }
+            break;
```

Comment:
We need to walk the parents to see if any are pinned (we would do this normally in `CInode::get_export_pin`). The code is trying to avoid adding inodes which are pinned by their parent (John's idea) and instead just add the parent.